cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +308,435 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
+      node_type.data() + level_boundaries[level],
+      thrust::make_zip_iterator(parent_col_id.begin() + level_boundaries[level - 1],
+                                scatter_indices.begin() + level_boundaries[level - 1]));
+    // Primary sort on parent_col_id
+    thrust::stable_sort_by_key(
+      rmm::exec_policy(stream),
+      parent_col_id.begin() + level_boundaries[level - 1],
+      parent_col_id.begin() + level_boundaries[level],
+      thrust::make_zip_iterator(node_type.data() + level_boundaries[level - 1],
+                                scatter_indices.begin() + level_boundaries[level - 1]));
+    PRINT_LEVEL_DATA(level, ".after sort");
+    nvtxRangePop();
+    nvtxRangePushA("transform");
+    auto start_it = thrust::make_zip_iterator(parent_col_id.begin() + level_boundaries[level - 1],
+                                              node_type.data() + level_boundaries[level - 1]);
+    auto adjacent_pair_it = thrust::make_zip_iterator(start_it - 1, start_it);
+    thrust::transform(rmm::exec_policy(stream),
+                      adjacent_pair_it + 1,
+                      adjacent_pair_it + level_boundaries[level] - level_boundaries[level - 1],
+                      col_id.data() + level_boundaries[level - 1] + 1,
+                      [] __device__(auto adjacent_pair) -> size_type {
+                        auto lhs = thrust::get<0>(adjacent_pair),
+                             rhs = thrust::get<1>(adjacent_pair);
+                        return lhs != rhs ? 1 : 0;
+                      });
+    nvtxRangePop();
+    nvtxRangePushA("scan");
+    // includes previous level last col_id to continue the index.
+    thrust::inclusive_scan(rmm::exec_policy(stream),
+                           col_id.data() + level_boundaries[level - 1],
+                           col_id.data() + level_boundaries[level] + (level != num_levels - 1),
+                           // +1 only for not-last-levels, for next level start col_id
+                           col_id.data() + level_boundaries[level - 1]);
+    PRINT_LEVEL_DATA(level, ".after scan");
+    // TODO scatter/gather to restore original order. (but scatter_indices is not zero based here)
+    thrust::sort_by_key(
+      rmm::exec_policy(stream),
+      scatter_indices.begin() + level_boundaries[level - 1],
+      scatter_indices.begin() + level_boundaries[level],
+      thrust::make_zip_iterator(col_id.begin() + level_boundaries[level - 1],
+                                parent_col_id.data() + level_boundaries[level - 1]));
+    PRINT_LEVEL_DATA(level, ".after restore order");
+    nvtxRangePop();
+  }
+  nvtxRangePop();
+  nvtxRangePushA("restore");
+  // restore original order of col_id., and used d_tree members
+  // TODO would scatter be faster than radix-sort here for 3 values?
+  thrust::sort_by_key(rmm::exec_policy(stream),
+                      scatter_indices.begin(),
+                      scatter_indices.end(),
+                      thrust::make_zip_iterator(
```

Comment:
This is where we could just scatter to `scatter_indices` indices, though, right?